### PR TITLE
Mixed dim refinement tests

### PR DESCRIPTION
**Pull request by cahaynes**
Added a couple fixes to the mixed dimension refinement and some unit testing.

### CODE REVIEW

**Comment by pbauman** [comment]
The dbg test died on fem_system_ex4 with

> Adaptive step 0: 
Assertion `!elem->has_dofs(sys_num) || elem->p_refinement_flag() == Elem::JUST_REFINED || elem->old_dof_object' failed.

**Comment by pbauman** [comment]
You can use `git rebase -i`, and put `reword` next to each commit whose message you want to edit, to redo the commit messages in b54bde0 and c15cf89.

**Comment by pbauman** [comment]
Instead of d788999, please just rebase. (You can use `git rebase -i` to drop the last commit from history.)

File: tests/mesh/mixed_dim_mesh_test.C [hunk 1]
```diff
@@ -357,5 +359,787 @@ public:
...
+          }
+      }
+      Elem* quad = _mesh->add_elem( new Quad4 );
+      unsigned int x=1,y=1,z=2;
+      quad->set_node(0) = _mesh->node_ptr(x+4*y    +16*z    );
+      quad->set_node(1) = _mesh->node_ptr(x+4*y    +16*z + 1);
+      quad->set_node(2) = _mesh->node_ptr(x+4*(y+1)+16*z + 1);
+      quad->set_node(3) = _mesh->node_ptr(x+4*(y+1)+16*z    );
+
+      // 2D elements will have subdomain id 0, this one will have 1
```
Comment (pbauman) [inline tests/mesh/mixed_dim_mesh_test.C]: You mean 3D elements will have subdomain id 0?

Comment (roystgnr) [inline tests/mesh/mixed_dim_mesh_test.C]: This looks like a copy and paste atavism.

**Comment by jwpeterson** [comment]
I will fix the following issues before merge, but for future reference:

1. Please create a branch for development rather than making commits directly on your local master branch.  This will help you organize your work better, and prevent any potential confusion between origin/master and upstream/master.
1. As @pbauman already mentioned, you should not have any merge commits on PR branches (i.e. do not run git pull with your branch checked out), except in very in specific situations, as it just muddies up the history unnecessarily.  Instead, run  `git rebase upstream/master` and fix any conflicts before creating the PR.  A rebase may also be in order even if you don't have any conflicts, in case you branched from a relatively old commit on master...
1. This is just a matter of personal preference, but your later commits don't have the proper author email information (cahaynes <cahaynes@Finicky>).  You can fix that with "git rebase -i" and marking the commits to be amended with "e" for edit, and then running
```
git commit --amend --author="Christopher Haynes <cahaynes@buffalo.edu>"
```
on each one.

**Comment by cahaynes** [comment]
@pbauman I'll take a closer look at which commit broke fem_system_ex4, as you suggested.  

@jwpeterson Thank you for the information, I'll make sure to keep it in mind moving forward.

File: tests/mesh/mixed_dim_mesh_test.C (hunk 1 above)
Comment (cahaynes) [inline tests/mesh/mixed_dim_mesh_test.C]: Yup, it was meant to be 3D.

**Comment by roystgnr** [comment]
@cahaynes - the breakage in fem_system_ex4 appears to be something going out of sync between processors.  If I run your branch (with the merge commit replaced by a rebase) with 3 processors I get that even more clearly: two processors hit the parallel_only() call in limit_underrefined_boundary at the same time the third is calling max(max_satisfied) in _smooth_flags().

**Comment by roystgnr** [comment]
Okay, this is my bug, you just triggered it.  Let me see if I can work out a fix for you to cherry-pick.

**Comment by roystgnr** [comment]
Try the flag synching fix in 577663818f0c2a7bea7b44298b8a326d99a9bd88.

**Comment by roystgnr** [comment]
More specifically, @cahaynes, once you've backed off the merge commit and rebased, you'll want to follow John's much-more-helpful-than-Roy instructions: https://github.com/libMesh/libmesh/commit/577663818f0c2a7bea7b44298b8a326d99a9bd88#commitcomment-14016914

**Comment by cahaynes** [comment]
@roystgnr - Thank you for the help, would've taken me much longer to catch that.

**Comment by pbauman** [comment]
Looks like there was some trouble with the rebase. I'll close this out cherry-pick off the commits and open  a new PR.

**Pull request closed without merging**